src/args/arg_matches.rs

```diff
@@ -1,282 +1,572 @@
     ///
     /// # Examples
     ///
-    /// ```no_run
+    /// ```rust
+    /// # use clap::{App, Arg};
+    /// let m = App::new("myprog")
+    ///     .arg(Arg::with_name("debug")
+    ///         .short("d")
+    ///         .multiple(true))
+    ///     .get_matches_from(vec![
+    ///         "myprog", "-d", "-d", "-d"
+    ///     ]);
+    ///
+    /// assert_eq!(m.occurrences_of("debug"), 3);
+    /// ```
+    ///
+    /// This next example shows that counts actual uses of the argument, not just `-`'s
+    ///
+    /// ```rust
+    /// # use clap::{App, Arg};
+    /// let m = App::new("myprog")
+    ///     .arg(Arg::with_name("debug")
+    ///         .short("d")
+    ///         .multiple(true))
+    ///     .arg(Arg::with_name("flag")
+    ///         .short("f"))
+    ///     .get_matches_from(vec![
+    ///         "myprog", "-ddfd"
+    ///     ]);
+    ///
+    /// assert_eq!(m.occurrences_of("debug"), 3);
+    /// assert_eq!(m.occurrences_of("flag"), 1);
+    /// ```
+    pub fn occurrences_of<S: AsRef<str>>(&self, name: S) -> u8 {
+        self.args.get(name.as_ref()).map_or(0, |a| a.occurs)
+    }
+
+    /// Because subcommands are essentially "sub-apps" they have their own `ArgMatches` as well.
+    /// This method returns the `ArgMatches` for a particular subcommand or None if the subcommand
+    /// wasn't present at runtime.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
     /// # use clap::{App, Arg, SubCommand};
-    /// # let app_matches = App::new("myapp")
-    /// #     .subcommand(SubCommand::with_name("test")).get_matches();
-    /// if let Some(matches) = app_matches.subcommand_matches("test") {
-    ///     // Use matches as normal
+    /// let app_m = App::new("myprog")
+    ///     .arg(Arg::with_name("debug")
+    ///         .short("d"))
+    ///     .subcommand(SubCommand::with_name("test")
+    ///         .arg(Arg::with_name("opt")
+    ///             .long("option")
+    ///             .takes_value(true)))
+    ///     .get_matches_from(vec![
+    ///         "myprog", "-d", "test", "--option", "val"
+    ///     ]);
+    ///
+    /// // Both parent commands, and child subcommands can have arguments present at the same times
+    /// assert!(app_m.is_present("debug"));
+    ///
+    /// // Get the subcommand's ArgMatches instance
+    /// if let Some(sub_m) = app_m.subcommand_matches("test") {
+    ///     // Use the struct like normal
+    ///     assert_eq!(sub_m.value_of("opt"), Some("val"));
     /// }
     /// ```
-    pub fn subcommand_matches<'na>(&self, name: &'na str) -> Option<&ArgMatches> {
-        if let Some(ref sc) = self.subcommand {
-            if sc.name != name {
-                return None;
-            }
-            return Some(&sc.matches);
-        }
-        None
+    pub fn subcommand_matches<S: AsRef<str>>(&self, name: S) -> Option<&ArgMatches<'a>> {
+        self.subcommand.as_ref().map(|s| if s.name == name.as_ref() { Some(&s.matches) } else { None } ).unwrap()
```

Comment:
I've gone back and forth about this. The `unwrap` always unwraps an `Some(Some(&ArgMatches))` or a `Some(None)` Because the problem is finding out if `subcommand` is `Some` or `None` is only half the problem, the other part is whether or not the names match, THEN we re-wrap the matches only.

Another way to do it is 

```rust
self.subcommand.iter().filter(|s| s.name == name.as_ref()).next().and_then(|s| Some(&s.matches))
```

But to me this looks just as messy, and unless you're used to Rust iterators and `Option` / `Result`s this looks like madness.

So I've settled on the slightly longer:

```rust
if let Some(ref s) = self.subcommand {
    if s.name == name.as_ref() { return Some(&s.matches); }
}
None
```

Which I think clearly shows intent without being too long.